File: .NET/Microsoft.Recognizers.Text/ModelFactory.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;

namespace Microsoft.Recognizers.Text
{
    internal class ModelFactory<TModelOptions> : Dictionary<(string culture, Type modelType), Func<TModelOptions, IModel>>
    {
        private const string FallbackCulture = Culture.English;

        private static ConcurrentDictionary<(string culture, Type modelType, string modelOptions), IModel> cache =
            new ConcurrentDictionary<(string culture, Type modelType, string modelOptions), IModel>();

        public T GetModel<T>(string culture, bool fallbackToDefaultCulture, TModelOptions options)
               where T : IModel
        {
            if (TryGetModel(culture, options, out T model))
            {
                return model;
            }
            else if (fallbackToDefaultCulture && TryGetModel(FallbackCulture, options, out model))
            {
                return model;
            }

            throw new ArgumentException($"Could not find Model with the specified configuration: {culture}, {typeof(T).ToString()}");
        }

        public new void Add((string culture, Type modelType) config, Func<TModelOptions, IModel> modelCreator)
        {
            base.Add(GenerateKey(config.culture, config.modelType), modelCreator);
        }

        public void InitializeModels(string targetCulture, TModelOptions options)
        {
            this.Keys
                .Where(key => string.IsNullOrEmpty(targetCulture) || key.culture.Equals(targetCulture))
                .ToList()
                .ForEach(key => this.InitializeModel(key.modelType, key.culture, options));
        }

        private static (string culture, Type modelType) GenerateKey(string culture, Type modelType)
        {
            return (culture.ToLowerInvariant(), modelType);
        }

        private void InitializeModel(Type modelType, string culture, TModelOptions options)
        {
            this.TryGetModel(modelType, culture, options, out IModel model);
        }

        private bool TryGetModel<T>(string culture, TModelOptions options, out T model)
                where T : IModel
        {
            var result = this.TryGetModel(typeof(T), culture, options, out IModel outModel);
            model = (T)outModel;
            return result;
        }

        private bool TryGetModel(Type modelType, string culture, TModelOptions options, out IModel model)
        {
            model = default(IModel);
            if (string.IsNullOrEmpty(culture))
            {
                return false;
            }

            culture = Culture.MapToNearestLanguage(culture);

            // Look in cache
            var cacheKey = (culture, modelType, options.ToString());
            if (cache.ContainsKey(cacheKey))
            {
                model = cache[cacheKey];
                return true;
            }

            // Use Factory to create instance
            var key = GenerateKey(culture, modelType);
            if (this.ContainsKey(key))
            {
                var factoryMethod = this[key];
                model = factoryMethod(options);

                // Store in cache
                cache[cacheKey] = model;
                return true;
            }

            return false;
        }
    }
}

File: Java/libraries/recognizers-text/src/main/java/com/microsoft/recognizers/text/ModelFactory.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.recognizers.text;

import com.microsoft.recognizers.text.utilities.StringUtility;

import java.lang.reflect.Type;
import java.util.HashMap;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import org.javatuples.Pair;
import org.javatuples.Triplet;

public class ModelFactory<TModelOptions> extends HashMap<Pair<String, Type>, Function<TModelOptions, IModel>> {

    // cacheKey: (string culture, Type modelType, string modelOptions)
    private static ConcurrentHashMap<Triplet<String, Type, String>, IModel> cache = new ConcurrentHashMap<Triplet<String, Type, String>, IModel>();

    private static final String fallbackCulture = Culture.English;

    public <T extends IModel> T getModel(Class<T> modelType, String culture, boolean fallbackToDefaultCulture, TModelOptions options) throws IllegalArgumentException {
        IModel model = this.getModel(modelType, culture, options);
        if (model != null) {
            return (T)model;
        }

        if (fallbackToDefaultCulture) {
            model = this.getModel(modelType, fallbackCulture, options);
            if (model != null) {
                return (T)model;
            }
        }

        throw new IllegalArgumentException(
                String.format("Could not find Model with the specified configuration: %s, %s", culture, modelType.getTypeName()));
    }

    private IModel getModel(Type modelType, String culture, TModelOptions options) {
        if (StringUtility.isNullOrEmpty(culture)) {
            return null;
        }

        // Look in cache
        Triplet<String, Type, String> cacheKey = new Triplet<>(culture.toLowerCase(), modelType, options.toString());
        if (cache.containsKey(cacheKey)) {
            return cache.get(cacheKey);
        }

        // Use Factory to create instance
        Pair<String, Type> key = generateKey(culture, modelType);
        if (this.containsKey(key)) {
            Function<TModelOptions, IModel> factoryMethod = this.get(key);
            IModel model = factoryMethod.apply(options);

            // Store in cache
            cache.put(cacheKey, model);
            return model;
        }

        return null;
    }

    public void initializeModels(String targetCulture, TModelOptions options) {
        this.keySet().stream()
                .filter(key -> StringUtility.isNullOrEmpty(targetCulture) || key.getValue0().equalsIgnoreCase(targetCulture))
                .forEach(key -> this.initializeModel(key.getValue1(), key.getValue0(), options));
    }

    private void initializeModel(Type modelType, String culture, TModelOptions options) {
        this.getModel(modelType, culture, options);
    }

    @Override
    public Function<TModelOptions, IModel> put(Pair<String, Type> config, Function<TModelOptions, IModel> modelCreator) {
        return super.put(
                generateKey(config.getValue0(), config.getValue1()),
                modelCreator);
    }

    private static Pair<String, Type> generateKey(String culture, Type modelType) {
        return new Pair<>(culture.toLowerCase(), modelType);
    }
}
